File: agent-driven-architecture-main/examples/typescript/src/blocks/artifact/tools.ts
// ── blocks/artifact/tools — record, request, confirm ───────────────────────
// `recordFinding` and `requestSeal` are Reversible: writing a draft line and
// asking to seal are both undoable. `confirmSeal` is Irreversible — it is the
// canonical irreversible action 14.3 names, and sealing is what 12.2's Drain
// performs. Its `requestedBy` reads the pending authority off the seal status's
// PARENT-declared property, so the gate compares principals without this file
// knowing anything about authority at all.

import type { Verb } from "@adr/spine/pure/verb";
import { irreversible, reversible } from "@adr/spine/pure/verb";
import { object, string } from "valibot";
import type {
  ConfirmSealCommand,
  ConfirmSealResult,
  RecordFindingCommand,
  RecordFindingResult,
  RequestSealCommand,
  RequestSealResult,
} from "./contract";
import type { ArtifactSlice } from "./slice";

const nothing = object({});

export function artifactVerbs<S>(read: (state: S) => ArtifactSlice): readonly Verb<S>[] {
  return [
    reversible<S, { text: string }, RecordFindingResult, RecordFindingCommand>({
      name: "recordFinding",
      describe: "Append one line to the session's work product.",
      schema: object({ text: string() }),
      run: (input) => ({ outcome: "ok", tool: "recordFinding", text: input.text }),
      sign: (result, sig, id) => ({
        outcome: "ok",
        tool: "recordFinding",
        sig,
        id,
        text: result.text,
      }),
    }),
    reversible<S, Record<string, never>, RequestSealResult, RequestSealCommand>({
      name: "requestSeal",
      describe:
        "Request that the work product be sealed and delivered. Reversible; delivers nothing.",
      schema: nothing,
      run: () => ({ outcome: "ok", tool: "requestSeal" }),
      sign: (_result, sig, id) => ({ outcome: "ok", tool: "requestSeal", sig, id }),
    }),
    irreversible<S, Record<string, never>, ConfirmSealResult, ConfirmSealCommand>({
      name: "confirmSeal",
      describe: "Confirm the seal. IRREVERSIBLE: this delivers the work product.",
      schema: nothing,
      run: () => ({ outcome: "ok", tool: "confirmSeal" }),
      sign: (_result, sig, id) => ({ outcome: "ok", tool: "confirmSeal", sig, id }),
      requestedBy: (state) => read(state).seal.requestedBy,
    }),
  ];
}

File: agent-driven-architecture-main/examples/kotlin/block/artifact/src/main/kotlin/adr/blocks/artifact/Tools.kt
// ── blocks/artifact/tools — the Verb table ─────────────────────────────────
//   recordFinding  Reversible    appends one line. NO effect.
//   requestSeal    Reversible    Draft → Sealing. NO effect — a request is reversible.
//   confirmSeal    Irreversible  Sealing → Sealed, plus EXACTLY ONE DeliverArtifact.
//
// Sealing the session is 14.3's canonical irreversible action, so it is gated by the
// same boundary check as paging on-call — the same mechanic, not a special case.

package adr.blocks.artifact

import adr.contract.ArtifactCommand
import adr.contract.ArtifactResult
import adr.spine.pure.Lens
import adr.spine.pure.RawInput
import adr.spine.pure.ToolName
import adr.spine.pure.Verb

val RECORD_FINDING = ToolName("recordFinding")
val REQUEST_SEAL = ToolName("requestSeal")
val CONFIRM_SEAL = ToolName("confirmSeal")

internal data class FindingInput(val text: String)

/**
 * The input of a verb that takes none. A CLASS, not a `data object`: an object has no
 * instantiation, which is the same defect as a top-level function, and this one is a
 * VALUE the decode step returns rather than a singleton anything depends on.
 */
internal class NoInput

internal class ArtifactTools<S>(private val lens: Lens<S, ArtifactSlice>) {

    fun verbs(): List<Verb<S, *, *>> = listOf(
        Verb.Reversible(
            name = RECORD_FINDING,
            describe = "Record one finding as a line in the session's work product.",
            decode = ::decodeFinding,
            run = { input, _ -> ArtifactResult.RecordFinding(RECORD_FINDING, input.text) },
            sign = { r, sig, id -> ArtifactCommand.RecordFinding(r.tool, sig, id, r.text) },
            narrow = { it as? ArtifactResult.RecordFinding },
        ),
        Verb.Reversible(
            name = REQUEST_SEAL,
            describe = "Request that the work product be sealed and delivered. Reversible.",
            decode = ::decodeNothing,
            run = { _, _ -> ArtifactResult.RequestSeal(REQUEST_SEAL) },
            sign = { r, sig, id -> ArtifactCommand.RequestSeal(r.tool, sig, id) },
            narrow = { it as? ArtifactResult.RequestSeal },
        ),
        Verb.Irreversible(
            name = CONFIRM_SEAL,
            describe = "Confirm the seal. IRREVERSIBLE: it delivers the work product.",
            decode = ::decodeNothing,
            run = { _, _ -> ArtifactResult.ConfirmSeal(CONFIRM_SEAL) },
            sign = { r, sig, id -> ArtifactCommand.ConfirmSeal(r.tool, sig, id) },
            narrow = { it as? ArtifactResult.ConfirmSeal },
            requestedBy = { state, _ ->
                (lens(state).seal as? SealStatus.Sealing)?.requestedBy
            },
        ),
    )

    private fun decodeFinding(raw: RawInput): FindingInput? =
        raw.text("text")?.let { FindingInput(it) }

    private fun decodeNothing(raw: RawInput): NoInput = NoInput()
}
